pkg/controller/drain.go

```diff
@@ -487,87 +493,110 @@ func filterSharedPVs(pvMap map[string][]string) {
 	glog.V(4).Info("Removed shared volumes. Filtered list of pods with volumes: ", pvMap)
 }
 
-func (o *DrainOptions) evictPodsWithPv(pods []*corev1.Pod,
+func (o *DrainOptions) evictPodsWithPv(attemptEvict bool, pods []*corev1.Pod,
 	policyGroupVersion string,
 	getPodFn func(namespace, name string) (*api.Pod, error),
 	returnCh chan error,
-	stopCh <-chan struct{},
 ) {
-	remainingPods := 0
-
 	sortPodsByPriority(pods)
 
 	volMap := o.doAccountingOfPvs(pods)
 
-EvictingPods:
-	for {
-		retryPods := []*corev1.Pod{}
-
-		for i, pod := range pods {
-			select {
-			case <-stopCh:
-				glog.Warningf("The caller function returned. No need to try evictions now")
-				return
-			default:
-			}
-
-			err := o.evictPod(*pod, policyGroupVersion)
-
-			if apierrors.IsTooManyRequests(err) {
-				// Pod eviction failed because of PDB violation, we will retry one we are done with this list.
-				glog.V(3).Info("Pod ", pod.Namespace, "/", pod.Name, " couldn't be evicted because of PDB violation. Will be retried.")
-				retryPods = append(retryPods, pod)
-				continue
-			} else if apierrors.IsNotFound(err) {
-				glog.V(3).Info("\t", pod.Name, " is already gone")
-				returnCh <- nil
-				continue
-			} else if err != nil {
-				returnCh <- fmt.Errorf("Error when evicting pod: %v/%v. Will not be retried. Err: %v", pod.Namespace, pod.Name, err)
-				continue
+	var (
+		remainingPods []*api.Pod
+		fastTrack     bool
+		nretries      = int(o.MaxEvictRetries)
+	)
+
+	if attemptEvict {
+		for i := 0; i < nretries; i++ {
+			remainingPods, fastTrack = o.evictPodsWithPVInternal(attemptEvict, pods, volMap, policyGroupVersion, getPodFn, returnCh)
+			if fastTrack || len(remainingPods) == 0 {
+				//Either all pods got evicted or we need to fast track the return (node deletion detected)
+				break
 			}
 
-			glog.V(3).Info("waiting for PVs to detach from node for pod: ", pod.Name)
-			// Eviction was successful. Wait for pvs for this pod to detach
-			pvs := volMap[pod.Namespace+"/"+pod.Name]
-			err = o.waitForDetach(pvs, o.nodeName, stopCh)
-
-			if apierrors.IsNotFound(err) {
-				glog.V(3).Info("Node not found anymore")
-				returnCh <- nil
-				remainingPods = len(pods) - (i + 1) + len(retryPods)
-				// The node itself is gone. Stop evicting pods.
-				// We should anyway return successful eviction for the remaining pods
-				break EvictingPods
-			} else if err != nil {
-				glog.Errorf("Error when waiting for volume to detach from node. Err: %v", err)
-				returnCh <- err
-				continue
-			}
-			glog.V(3).Info("Detached pv for pod: ", pod.Name)
-			returnCh <- nil
+			glog.V(4).Info("Eviction for some pods will be retried after ", PodEvictionRetryInterval)
```

Comment:
Minor: In case when ntries are exhausted, we will go and try a delete. In that case this log will be misleading.